### PR TITLE
algorithm arguments fix (towards #1618)

### PR DESCRIPTION
**Pull request by rupertford**
The new approach to parsing the algorithm layer using the PSyIR did not support negative literal values as arguments. This PR fixes this.

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review from @arporter or @sergisiso

File: src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py [hunk 1]
```diff
@@ -191,8 +192,8 @@ def test_ai2psycall_apply_error(fortran_reader):
     with pytest.raises(TypeError) as info:
         trans.apply(invoke)
     assert ("Expected Algorithm-layer kernel arguments to be a literal, "
```
Comment (arporter) [inline src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py:194]: Oh, I see that we explicitly don't support expressions (I'd forgotten that) as arguments so you can disregard my previous comment.

File: src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py [hunk 2]
```diff
@@ -466,19 +467,35 @@ def test_ai2psycall_apply_invoke_symbols_scope(fortran_reader):
...
+           "reference or code block, but '+1.0' is of type 'UnaryOperation'."
+           in str(info.value))
+
+    # Invalid argument exception (Node parent is a KernelFunctor)
+    _ = KernelFunctor.create(
+        DataTypeSymbol("my_kernel", REAL_TYPE), [arg])
+    with pytest.raises(TypeError) as info:
+        AlgInvoke2PSyCallTrans._add_arg(arg, [])
+    assert("Expected Algorithm-layer kernel arguments to be a literal, "
+           "reference or code block, but '+1.0' in kernel 'my_kernel' is "
```
Comment (arporter) [inline src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py:492]: s/in/passed to/

File: src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py [hunk 3]
```diff
@@ -130,10 +134,16 @@ def _add_arg(arg, arguments):
...
+                if arg.parent and isinstance(arg.parent, KernelFunctor):
+                    string = f"{string} in kernel '{arg.parent.name}'"
+                string = f"{string} is of type '{type(arg).__name__}'."
+            else:
+                string = f"but found '{type(arg).__name__}'."
             raise TypeError(
                 f"Expected Algorithm-layer kernel arguments to be "
-                f"a literal, reference or code block, but "
-                f"found '{type(arg).__name__}'.")
+                f"a literal, reference or code block, {string}.")
```
Comment (arporter) [inline src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py:146]: Probably better to capitalise these to emphasise that they are Node types (i.e. "Literal, Reference or CodeBlock").

**Review by arporter — CHANGES_REQUESTED** [review]
Nearly perfect. I think the error message could be improved slightly but that's all. Presumably we are also getting closer to being able to support expressions in kernel arguments?
Tests and examples are all fine with compilation.
Changes are pylint clean.
This a bug fix/towards PR so no doc changes needed.

File: src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py [hunk 4]
```diff
@@ -118,7 +119,10 @@ def _add_arg(arg, arguments):
         '''
         sym_maths = SymbolicMaths.get()
 
-        if isinstance(arg, Literal):
+        if isinstance(arg, Literal) or (
+                isinstance(arg, UnaryOperation) and
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py:123]: As you point out below, I've continued with the current restrictions that we have on arguments. We are close to being able to support expressions if the Met Office decide they want that. The only problem (that I know about) is determining the datatype and precision of the argument if it is an expression.

File: src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py:492]: Better, thanks.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Pull request merged**